HCP HashiStack Use Case (#4487)
hashicorp/waypoint: website/content/docs/use-cases/hcp-hashistack-quickstart.mdx

---
layout: docs

> page_title: A Complete Cloud Ecosystem in Two Commands with Waypoint, Terraform, and the HashiStack

Reviewer: Feels very wordy but it gets the point across?
```suggestion
page_title: A Complete Cloud Ecosystem built with Terraform, HCP Waypoint, HCP Vault, HCP Consul, and Nomad
```

then launch the application to an EC2 container within our secured internal network with `waypoint up`.
Welcome to the HashiCorp stack magic.

The ingredients for this special dish are [Waypoint](https://www.waypointproject.io/),
[Terraform](https://www.terraform.io/), [Consul](https://www.consul.io/), [Nomad](https://www.nomadproject.io/), and
[Vault](https://www.vaultproject.io/).
We chose to use HashiCorp managed versions of Waypoint, Consul, and Vault for one convenient HCP UI and
to showcase the [HashiCorp Virtual Network (HVN)](https://developer.hashicorp.com/hcp/docs/hcp/network).
The same experience is of course possible with self-managed versions.

## Prerequisites

To follow along with this use case on your own machine, you will need:

- A [HashiCorp Cloud Platform (HCP)](https://portal.cloud.hashicorp.com/sign-in) account.
- An [AWS account](https://aws.amazon.com/account/) with AWS Access Credentials configured locally:

```shell
export AWS_ACCESS_KEY_ID=<your AWS access key ID>
export AWS_SECRET_ACCESS_KEY=<your AWS secret access key>
export AWS_SESSION_TOKEN=<your AWS session token>
```

- Docker registry

> - Credentials to push to the docker registry

Reviewer 2: ```suggestion
- Credentials to push to the Docker registry
```

write configuration files for [Terraform](https://developer.hashicorp.com/terraform/language) and [Waypoint](https://developer.hashicorp.com/waypoint/docs/waypoint-hcl/syntax).

In this use case, we will configure our infrastructure as code, which Terraform will run to automatically provision the following:

- [AWS Virtual Private Cloud](https://aws.amazon.com/pt/vpc/) and peer it with our [HashiCorp Virtual Network (HVN)](https://developer.hashicorp.com/hcp/docs/hcp/network)
- [HCP Consul](https://cloud.hashicorp.com/products/consul) service mesh to manage services within the secure private network
- [HCP Vault](https://cloud.hashicorp.com/products/vault) cluster to securely store credentials
- [Nomad](https://developer.hashicorp.com/nomad/tutorials/get-started/get-started-intro) cluster on AWS EC2 instance to manage application containers.

We will write application deployment configurations as code, which [HCP Waypoint](https://cloud.hashicorp.com/products/waypoint)
will run to build the container image and deploy the application to the infrastructure we set up with Terraform.

All it takes are two commands:

> `terraform apply`
> `waypoint up`

Reviewer 2: ```suggestion
- `terraform apply`
- `waypoint up`
```

Might be nice to have these as a list in markdown so they look good formatted!

![HCP Service Principal](/img/use-cases/hcp-hashistack-quickstart/service-principal.png)

When client credentials are set, they are always used by the HCP Provider client, regardless of an existing user session.

## Define the Virtual Network

### VPC

You can find VPC configurations in `vpc.tf`. There is more than one way to create this VPC, including manually in AWS.
The Amazon VPC can be managed with the [AWS provider](https://registry.terraform.io/providers/hashicorp/aws/latest/docs).

```terraform
provider "aws" {
   region = var.region
}

module "vpc" {
  source = "terraform-aws-modules/vpc/aws"
  name   = var.name
  cidr   = var.cidr
  public_subnets = var.public_subnets
  public_subnet_tags = {
    Name = "${var.name}-public"
  }

  azs            = slice(data.aws_availability_zones.available.*.names[0], 0, 3)
   enable_dns_hostnames            = true
  enable_ipv6                     = true
  assign_ipv6_address_on_creation = true
  public_subnet_ipv6_prefixes     = [0, 1, 2]

  tags = {
    Terraform = "true"
    Name      = var.name
  }
}

data "aws_availability_zones" "available" {
  state = "available"
}
```

### HVN

> In the [`hcp.tf`](https://github.com/xiaolin-ninja/2048_hcp_hashistack/blob/main/hcp.tf) configuration file,

Reviewer 2: For links like this, I wonder if it makes more sense to link directly to the latest commit sha rather than `main`? If that file ever gets moved, we'll have dead links in the guide. Linking to a specific commit in the source tree would at least keep it active.

  cloud_provider = "aws"
  region         = var.region
}

resource "hcp_aws_network_peering" "peer" {
  hvn_id          = hcp_hvn.xx_2048_hvn.hvn_id
  peering_id      = "hcp-hashistack"
  peer_vpc_id     = module.vpc.vpc_id
  peer_account_id = module.vpc.vpc_owner_id
  peer_vpc_region = var.region
}

resource "hcp_hvn_route" "hvn-to-vpc" {
  hvn_link         = hcp_hvn.xx_2048_hvn.self_link
  hvn_route_id     = "hvn-to-vpc"
  destination_cidr = module.vpc.vpc_cidr_block
  target_link      = hcp_aws_network_peering.peer.self_link
}

resource "aws_vpc_peering_connection_accepter" "peer" {
  vpc_peering_connection_id = hcp_aws_network_peering.peer.provider_peering_id
  auto_accept               = true
  accepter {
    allow_remote_vpc_dns_resolution = true
  }
}

resource "aws_route" "hvn-peering" {
  route_table_id            = module.vpc.public_route_table_ids[0]
  destination_cidr_block    = hcp_hvn.xx_2048_hvn.cidr_block
  vpc_peering_connection_id = aws_vpc_peering_connection_accepter.peer.id
```

## Define HCP Consul

HCP Consul enables you to quickly deploy Consul servers in AWS while offloading the operations burden to the SRE experts at HashiCorp.

> Provisioning HCP Consul is incredibly easy.

Reviewer: I think to fit our documentation guidelines (avoiding "just," "easily," etc), it would be better to phrase like:
```suggestion
HashiCorp has worked to make provisioning HCP Consul as easy as possible.
```

You can define the cluster in Terraform using the [`hcp_consul_cluster`](https://registry.terraform.io/providers/hashicorp/hcp/latest/docs/resources/consul_cluster) resource,
as shown in `hcp.tf`:

```terraform
resource "hcp_consul_cluster" "xx_2048_consul" {
  hvn_id          = hcp_hvn.xx_2048_hvn.hvn_id
  cluster_id      = "xx-2048-consul"
  tier            = "development"
  public_endpoint = true
}
```

> You can also provision a Consul cluster with one click in the HCP UI.

Reviewer: A screenshot here would be great!

Reviewer 2: 👍🏻 to this, as I'm reading through this, I think it would be nice to include some screenshots of things to help folks follow along

## Define HCP Vault

> Defining a Vault cluster in Terraform is as simple as with Consul,

Reviewer: Unsure how to word this, but trying to avoid the "simple" language:
```suggestion
Defining an HCP Vault cluster via Terraform also requires minimal configuration,
```

```terraform
resource "hcp_vault_cluster" "xx_2048_vault" {
  hvn_id          = hcp_hvn.xx_2048_hvn.hvn_id
  cluster_id      = "xx-2048-vault"
  public_endpoint = true
}
```

### Vault Cluster Configurations

We will further configure Vault to communicate with Nomad.
The [Vault provider](https://registry.terraform.io/providers/hashicorp/vault/latest/docs) allows Terraform to read from, write to, and configure HashiCorp Vault:

```terraform
provider "vault" {
  address   = hcp_vault_cluster.xx_2048_vault.vault_public_endpoint_url
  namespace = "admin"
  token     = hcp_vault_cluster_admin_token.xx_2048_vault_token.token
}
```

You can find the detailed Vault configurations in [`vault-config.tf`](https://github.com/xiaolin-ninja/2048_hcp_hashistack/blob/main/vault-config.tf)

## Define Nomad Configurations

In [`nomad.tf`](https://github.com/xiaolin-ninja/2048_hcp_hashistack/blob/main/nomad.tf),
we configure the AWS Security Group, find a demo AMI, and configure the Nomad server, client, and auto-scaling.
The scripts `user-data-client.sh` and `user-data-server.sh` are run directly in the EC2 instances spun up by Terraform to provision Nomad.

> ~> **Note:** Since this is a use-case demo, I am not following Security best-practices.

Reviewer: Nice note

For example, my AWS Security Group allows all ingress and egress traffic.
I am also provisioning only one Nomad server and client cluster each, which does not insulate against downtime.

> For real production use cases, please properly configure your security group and have multiple clusters.

Reviewer 2: I'd recommend re-wording this to remove `I` and be more generic. There might be other places where this needs to happen too. These use-cases won't have authors attached on our website, so the voicing doesn't make sense.

> This is how the launch template looks like:

Reviewer: active vs passive
```suggestion
The launch template looks like this:
```

## Deploy the Infrastructure!

Once you have finished defining configurations (or cloned them from the [repo](https://github.com/xiaolin-ninja/2048_hcp_hashistack)),
and set the environment variables for HCP and AWS, you are now ready to deploy your infrastructure!

```shell
$ terraform init
Initializing the backend...
Initializing provider plugins...
...
Terraform has been successfully initialized!
...
```

Once Terraform has been initialized, you can verify that the resources will
be created using the `plan` command.

```shell
$ terraform plan
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create
 <= read (data resources)
Terraform will perform the following actions:
...
...
Plan: 11 to add, 0 to change, 0 to destroy.
```

Finally, this is the magic moment I promised:

```shell
$ terraform apply
...
Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.
  Enter a value:
```

Remember to confirm the run by entering `yes`!

## Set up HCP Waypoint Context

> HCP Waypoint is currently in Beta, so we don't have a convenient Terraform module to provision Waypoint automatically yet (as of February 2023).

Reviewer: I think it's fine to say this much about that work, but open to feedback.
```suggestion
HCP Waypoint is currently in Beta, and a Terraform module to provision Waypoint is still under development.
```

![Waypoint Install New Runner](img/use-cases/hcp-hashistack-quickstart/waypoint-runner-install.png)

You will need to export the Nomad server address, which you can find on the AWS EC2 instance:
![Nomad EC2 Instances](img/use-cases/hcp-hashistack-quickstart/nomad-ec2.png)

```shell
export NOMAD_ADDR=<nomad_private_addr>
```

![Waypoint Install New Runner CLI](img/use-cases/hcp-hashistack-quickstart/wp-runner-install.png)

Export the nomad address to your local environment and install the runner after installing the HCP Waypoint context,
and you're all set to use HCP Waypoint.

The `waypoint.hcl` file defines application deployment configurations.
For this use case, we have a simple "build", "deploy", and "release" flow:

```hcl
project = "2048-hashistack"

app "nomad" {
  build {
    use "docker" {}
    registry {
      use "docker" {
        ...
      }
    }
  }

  deploy {
    use "nomad" {
      datacenter = "hcp-2048"
    }
  }
}
```

You can [read more about Waypoint HCL](https://developer.hashicorp.com/waypoint/docs/waypoint-hcl) fundamentals,
or explore [advanced deployment workflows using custom pipelines](https://developer.hashicorp.com/waypoint/docs/pipelines).

> Now, run `waypoint init`:

Reviewer: We should come back and update this to point to the WP Terraform Provider once Clint has migrated it!
https://github.com/hashicorp-dev-advocates/terraform-provider-waypoint

```shell
$ terraform plan
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create
 <= read (data resources)
Terraform will perform the following actions:
...
...
Plan: 11 to add, 0 to change, 0 to destroy.
```

Finally, this is the magic moment we promised:

```shell
$ terraform apply
...
Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.
  Enter a value:
```

Remember to confirm the run by entering `yes`!

## Set up HCP Waypoint Context

HCP Waypoint is currently in Beta, and a Terraform module to provision Waypoint is still under development.

Head into your HCP UI and [follow this tutorial](/waypoint/tutorials/runners/hcp-runners-on-your-infra#set-up-the-environment)
to set up your local environment.

Navigate to Waypoint in the HCP dashboard:
![HCP UI](img/use-cases/hcp-hashistack-quickstart/hcp.png)

Then click on "Install New Runner", which will generate terminal inputs to set up the Waypoint context and install a Runner on your chosen platform.
In this case, we choose Nomad with a host volume named "wp_runner".

> ![Waypoint Install New Runner](img/use-cases/hcp-hashistack-quickstart/waypoint-runner-install.png)

Reviewer 3: I could not see this or the subsequent 7 images in Vercel 🤔 I think that there needs to be a `/` preceding `img`.
<img width="715" alt="Screen Shot 2023-02-08 at 12 01 59 PM" src="https://user-images.githubusercontent.com/83741749/217600108-8d9f46f2-6727-412b-a96d-2c108d24b7ec.png">

Author: Oh good catch, I assumed it was a rendering issue like with the images in Github markdown preview.
